engine/server/engine/enclave_manager/enclave_pool.go

```diff
@@ -203,34 +242,58 @@ func (pool *EnclavePool) destroyPreviousIdleEnclaves() error {
 }
 
 // run method is in charge of filling the enclave pool channel with new idle enclaves
-func (pool *EnclavePool) run() {
+func (pool *EnclavePool) run(ctx context.Context) {
 
-	defer close(pool.idleEnclavesChan)
+	defer func() {
+		close(pool.idleEnclavesChan)
+	}()
 
-	fails := 0
+	// init the pool
+	if err := pool.fill(ctx); err != nil {
+		logrus.Errorf("An error occurred filling the enclave pool. Error\n%v", err)
+		return
+	}
 
 	for {
-		newEnclaveInfo, err := pool.createNewEnclave()
-		if err != nil {
-			// Retry strategy if something fails
-			logrus.Errorf("An error occurred creating a new idle enclave. Error\n%v", err)
-			fails++
-			if fails >= maxRetryAfterFails {
-				logrus.Debugf("The enclave pool sub-routine has been canceled after reaching the max retries value '%v'", maxRetryAfterFails)
+		// wait until receive the re-fill signal or the close signal
+		select {
+		case <-pool.fillChan:
+			if err := pool.fill(ctx); err != nil {
+				if err == context.Canceled {
+					logrus.Debug("The sub-routine context has been canceled")
+				} else {
+					logrus.Errorf("An error occurred filling the enclave pool. Error\n%v", err)
+				}
 				break
 			}
-			logrus.Debugf("Next retry will be executed on '%v' seconds", timeBetweenRetries.Seconds())
-			time.Sleep(timeBetweenRetries)
+		case <-pool.closeChan:
+			return
+		}
+	}
+}
+
+func (pool *EnclavePool) fill(ctx context.Context) error {
+
+	for {
+		if len(pool.idleEnclavesChan) == cap(pool.idleEnclavesChan) {
```

Comment:
Huh... You'll get some concurrency issues doing that here. What happens if an enclave get picked WHILE it's being filled? With this for loop checking on the channel length, you're basically defeating the purpose of using channels, aren't you?

I think another approach here could be to have this function create just ONE enclave, and then you use a fillChan buffered channel that you start by filling entirely. This way, when you start the subroutine, it will start by calling X number of time this function to empty the fillChan completely, and continue from there. If you do that, then it makes total sense to use a buffered channel for fillChan

Reply:
refactored! good suggestion!

Reply:
in this version, the race condition was handled by the Select statement in the run method, because it will be executing one call to fill at a time, but your suggestion I think is more reasonable and legible, so I'll do the refactor.